### PR TITLE
react-hooks: Add usePresence listener hook

### PR DESCRIPTION
**Pull request by splindsay-92**
### Context

* [CHA-388]

### Description

- Added the `usePresenceListener` hook to the Chat React library. This allows the user to interact with the underlying `Presence` instance, and subscribe a listener to presence events.
- The hook also returns the current state of presence members, which is kept up to date by the hook.

- Updated the demo app to use the new hook, the presence list is now updated via the returned state from the hook.
- Updated the readme to document hook usage.
- Added unit tests for the new hook.
- Added integration tests for the new hook.

### Checklist

* [x] QA'd by the author.
* [ ] Unit tests created (if applicable).
* [x] Integration tests created (if applicable).
* [x] Follow coding style guidelines found [here](https://github.com/ably/engineering/tree/main/best-practices).
* [x] TypeDoc updated (if applicable).
* [x] (Optional) Update documentation for new features.
* [x] Browser tests created (if applicable).
* [x] In repo demo app updated (if applicable).

### Testing Instructions (Optional)

1. Spin up a local instance of the demo app.
2. Open up two tabs with different clients and join presence
3. Ensure both tabs show presence for both clients

[CHA-388]: https://ably.atlassian.net/browse/CHA-388?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/react/hooks/use-presence-listener.tsx
Comment (AndyTWF) [inline src/react/hooks/use-presence-listener.tsx:87]: Eventually this might be useful functionality to have in presence itself (so you subscribe to the whole presence set), but appreciate that's not the priority right now.

Could you add a ticket to the backlog to that effect?

File: src/react/hooks/use-presence-listener.tsx
Comment (AndyTWF) [inline src/react/hooks/use-presence-listener.tsx:117]: It would be good, at some point, to have our hooks write to the actual logger in the `ChatClient`.

We should do a hook (internal only) that does `useLogger` or similar, to access the chat clients logger.

Can we create a ticket to look into that post release?

**Review by AndyTWF — COMMENTED** [review]
A few comments - I won't approve as I'll be away when it gets merged.

Also tests + README changes

File: src/react/hooks/use-presence-listener.ts
Comment (vladvelici) [inline src/react/hooks/use-presence-listener.ts:90]: do we want all this retry stuff in react only or do we think it's useful in the library too?

Comment (splindsay-92) [reply src/react/hooks/use-presence-listener.ts:90]: Andy mentioned this [above](https://github.com/ably/ably-chat-js/pull/324#discussion_r1720037282), I've added a ticket to go back and this to the core SDK :)

File: demo/src/components/UserPresenceComponent/UserPresenceComponent.tsx
Comment (AndyTWF) [inline demo/src/components/UserPresenceComponent/UserPresenceComponent.tsx:8]: Can we re-add the line between imports and interface here, for clarity

File: src/react/README.md
Comment (AndyTWF) [inline src/react/README.md:198]:
```suggestion
It is intended solely for monitoring the state of `Presence` in a room, should you wish to enter presence and thus update the presence state,
```

File: test/react/hooks/use-presence-listener.test.tsx
Comment (AndyTWF) [inline test/react/hooks/use-presence-listener.test.tsx:49]: Remove console log

**Review by AndyTWF — APPROVED** [review]

File: src/react/hooks/use-presence-listener.ts
Comment (AndyTWF) [inline src/react/hooks/use-presence-listener.ts:76]: Might better to use `ReturnType<typeof setTimeout>` here (sorry it got lost on my review!)

**Pull request merged**
